documentation/optee_design.md

```diff
@@ -523,8 +523,104 @@ implementation, as explained in [crypto.md].
 TBD
 
 # 12. Trusted Applications
-## Format
-Trusted Applications consists of a signed ELF file.
+
+## Static and Dynamic TAs
+
+There are two ways to implement Trusted Applications, static and dynamic.
+For most cases, Dynamically-loaded TAs are preferred.
+
+### Static / Pseudo Trusted Applications
+
+These are added directly to the optee_os tree in, eg, `core/arch/arm/sta`, and
+are built along with and statically built into the OS blob.
+
+The Static / Pseudo Trusted Applications included in OP-TEE already are OP-TEE
+core secure privilege layer services hidden behind a "GP TA Client" API; it's
+in these "Pseudo Trusted Applications" that the core functionality for
+communication with the REE (like "Open Session"), and API handlers from Dynamic
+Trusted Applications are implemented.
+
+Static TAs do not benefit from the 'GP Core Internal API' support
+specified by the GP TEE specs. Static TAs can only use the OP-TEE core internal
+APIs and routines.
+
+Pseudo-TAs have the same privileged view of Secure World as the optee_os core
+itself.  For complex TAs, that may not be desirable.
+
+In most cases a real, dynamically loaded TA is the best choice instead of adding
+your code directly to the OP-TEE core.  However if you decide your application
+is best handled directly in the core like this, you can look at
+`core/arch/arm/sta/stats.c` as a template and just add your static TA based on
+that to the `sub.mk` in the same directory.
+
+###Trusted Applications
+
+Trusted Applications (TAs) are applications dynamically loaded by OP-TEE
+core in the Secure World when something in the REE wants to talk to that
+particular application UUID.  It's similar to the way the Linux
+kernel can dynamically load kernel modules.
+
+Because the TAs are signed by the same key that built the OP-TEE core, they are
+able to be stored in the untrusted REE filesystem, and tee-supplicant will
+take care of passing them to be checked and loaded by the Secure World OP-TEE
+core.  Again this is simular to Linux kernel module signature checking.
+
+Trusted Application benefit from the `GP Core Internal API` as specified by
+the 'GP Specs'.
+
+Trusted Application consist of a cleartext signed ELF file, named from the UUID
+of the TA and the suffix ".ta".
+
+They are built separately from the optee_os boot-time blob, although when they
+are built they use the same build system, and are signed with the key from the
+build of the original optee_os blob.
+
+## Special treatment of Trusted Applications
+
+### Syscalls
+
+Since (dynamically loaded) TAs are not directly bound to function exports in
+the optee_os blob, since that would require upgrades of all TAs synchronusly to
+upgrades of the optee_os blob, the use of optee_os exports by the TA must be
+resolved at runtime.
```

Comment:
The fundamental reason for having syscalls is that we're switching privilege / exception level. So that's not just a matter of symbol resolution (dynamic vs. static linking), as it is the case for kernel modules.